android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/BaseViewRouter.kt

```diff
@@ -0,0 +1,160 @@
+    private var ribId: Int? = null
+    internal var view: V? = null
+    protected var parentViewGroup: ViewGroup? = null
+
+    private lateinit var attachPermanentParts: RoutingAction<V>
+    protected open val permanentParts: List<() -> BaseViewRouter<*, *>> = emptyList()
+    private var savedViewState: SparseArray<Parcelable> = SparseArray()
+
+    private fun attachPermanentParts() {
+        permanentParts.forEach {
+            addChild(it()) // fixme save and restore these as well
+        }
+    }
+
+    override fun dispatchAttach(savedInstanceState: Bundle?, tag: String) {
+        super.dispatchAttach(savedInstanceState, tag)
+        updateRibId(savedInstanceState?.getInt(KEY_RIB_ID, generateGroupId()) ?: generateGroupId())
+        savedViewState = savedInstanceState?.getSparseParcelableArray<Parcelable>(KEY_VIEW_STATE) ?: SparseArray()
+        attachPermanentParts()
+    }
+
+    private fun generateGroupId(): Int =
+        requestCodeRegistry.generateGroupId(ribName)
+
+    private fun generateRequestCode(code: Int): Int =
+        requestCodeRegistry.generateRequestCode(ribName, code)
+
+    private fun updateRibId(value: Int) {
+        ribId = value
+    }
+
+    override fun saveInstanceState(outState: Bundle) {
+        super.saveInstanceState(outState)
+        outState.putInt(KEY_RIB_ID, ribId ?: generateGroupId().also { updateRibId(it) })
+        outState.putSparseParcelableArray(KEY_VIEW_STATE, savedViewState)
+    }
+
+    // todo move this to base router class so no casting is needed inside foreach
+    fun onAttachToView(parentViewGroup: ViewGroup) {
+        this.parentViewGroup = parentViewGroup
+        view = createView(parentViewGroup)
+        view?.let {
+            it.androidView.restoreHierarchyState(savedViewState)
+            parentViewGroup.addView(it.androidView)
+            (interactor as BaseInteractor<*, *>).onViewCreated() // FIXME merge BaseInteractor to Interactor
+        }
+
+        children.forEach {
+            attachChildToView(it as BaseViewRouter<*, *>)
+        }
+    }
+
+    private fun createView(parentViewGroup: ViewGroup): V? =
+        viewFactory?.invoke(parentViewGroup)
+
+    fun addChild(child: BaseViewRouter<*, *>, bundle: Bundle? = null) {
+        attachChildToView(child)
+        // todo refactor base implementation so that this branching is not necessary
+        if (bundle != null) {
+            attachChild(child, bundle)
+        } else {
+            attachChild(child)
+        }
+    }
+
+    protected fun attachChildToView(child: BaseViewRouter<*, *>) {
+        parentViewGroup?.let {
+            child.onAttachToView(
+                attachTargetForConfiguration(view, child, it)
+            )
+        }
+    }
+
+    open fun attachTargetForConfiguration(view: V?, child: BaseViewRouter<*, *>, parentViewGroup: ViewGroup): ViewGroup =
+        view?.androidView ?: parentViewGroup
```

Comment:
Is it possible to have branching outside of this method?
Otherwise, if you want user to be able to choose where he wants to attach the view, should we reorder the parameters?
`view, parentView, router`

Reply:
Just realised this should be named attachTargetForChild, not Configuration. And then it makes sense in this order: childRouter, view, parentViewGroup, what do you think?